qcfractal/interface/collections/dataset.py

```diff
@@ -942,52 +1075,91 @@ def add_contributed_values(self, contrib: ContributedValues, overwrite=False) ->
+                spec["stoich"] = "default"
+            if isinstance(theory_level_details, dict):
+                spec.update(**theory_level_details)
+
+            # check is search matches CV column
+            for key, value in search.items():
+                if value is None:
+                    continue
+                elif isinstance(value, str):
+                    value = value.lower()
+                    if not (key in spec and value == spec[key].lower()):
+                        break
+                elif isinstance(value, (list, tuple)):
+                    query = [x.lower() for x in value]
+                    if not (key in spec and spec[key].lower() in query):
+                        break
+                else:
+                    raise TypeError(f"Search type {type(value)} not understood.")
             else:
-                values = {k: [np.array(v)] for k, v in data.values.items()}
+                ret = ret.append(spec, ignore_index=True)
 
-        tmp_idx = pd.DataFrame.from_dict(values, orient="index", columns=[data.name])
+        if pretty:
+            ret.fillna("None", inplace=True)
+        ret.set_index(list(self.data.history_keys[:-1]), inplace=True)
+        ret.sort_index(inplace=True)
+        return ret
 
-        # Convert to numeric
-        tmp_idx[tmp_idx.select_dtypes(include=['number']).columns] *= constants.conversion_factor(
-            data.units, self.units)
+    def _get_contributed_values(self, **spec) -> pd.DataFrame:
+        queries = self._list_contributed_values(**spec).reset_index().to_dict("records")
+        ret = pd.DataFrame({"index": self.get_index()})
+        for query in queries:
+            data = self.data.contributed_values[query["name"].lower()].copy()
 
-        return tmp_idx
+            # Annoying work around to prevent some pands magic
+            if isinstance(next(iter(data.values.values())), (int, float)):
+                values = data.values
+            else:
+                # TODO temporary patch until msgpack collections
+                if self.data.default_driver == "gradient":
+                    values = {k: np.array(v).reshape(-1, 3) for k, v in data.values.items()}
+                else:
+                    values = {k: np.array(v) for k, v in data.values.items()}
+            column_name = data.name
+            ret[column_name] = list(values.values())
+            # Convert to numeric
+            try:
+                ret[column_name] *= constants.conversion_factor(
+                    data.units, self.units)
+            except ValueError as e:
+                # This is meant to catch pint.errors.DimensionalityError without importing pint, which is too slow
+                if e.__class__.__name__ == "DimensionalityError":
+                    ret.rename(columns={column_name: f"{column_name} [{data.units}]"}, inplace=True)
+                else:
+                    raise
+        ret.set_index("index", inplace=True)
```

Comment:
So here we will not change the underlying `ds.df`? In this case it appears we do not add the metadata to `self._column_metadata[name] = query` or similar. This could make the `visualization` function difficult.

Reply:
Added caching to ds.df and column_metadata for contributed values.